xds/src/main/java/io/grpc/xds/internal/sds/DynamicSslContextProvider.java

```diff
@@ -0,0 +1,134 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds.internal.sds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ImmutableList;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CertificateValidationContext;
+import io.envoyproxy.envoy.extensions.transport_sockets.tls.v3.CommonTlsContext;
+import io.grpc.Status;
+import io.grpc.xds.EnvoyServerProtoData.BaseTlsContext;
+import io.netty.handler.ssl.ApplicationProtocolConfig;
+import io.netty.handler.ssl.SslContext;
+import io.netty.handler.ssl.SslContextBuilder;
+import java.io.IOException;
+import java.security.cert.CertStoreException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/** Base class for dynamic {@link SslContextProvider}s. */
+public abstract class DynamicSslContextProvider extends SslContextProvider {
+
+  protected final List<Callback> pendingCallbacks = new ArrayList<>();
+  @Nullable protected final CertificateValidationContext staticCertificateValidationContext;
+  @Nullable protected SslContext sslContext;
+
+  protected DynamicSslContextProvider(
+      BaseTlsContext tlsContext, CertificateValidationContext staticCertValidationContext) {
+    super(tlsContext);
+    this.staticCertificateValidationContext = staticCertValidationContext;
+  }
+
+  @Nullable
+  public SslContext getSslContext() {
+    return sslContext;
+  }
+
+  protected abstract CertificateValidationContext generateCertificateValidationContext();
+
+  /** Gets a server or client side SslContextBuilder. */
+  protected abstract SslContextBuilder getSslContextBuilder(
+          CertificateValidationContext certificateValidationContext)
+      throws CertificateException, IOException, CertStoreException;
+
+  // this gets called only when requested secrets are ready...
+  protected final void updateSslContext() {
+    try {
+      CertificateValidationContext localCertValidationContext =
+          generateCertificateValidationContext();
+      SslContextBuilder sslContextBuilder = getSslContextBuilder(localCertValidationContext);
+      CommonTlsContext commonTlsContext = getCommonTlsContext();
+      if (commonTlsContext != null && commonTlsContext.getAlpnProtocolsCount() > 0) {
+        List<String> alpnList = commonTlsContext.getAlpnProtocolsList();
+        ApplicationProtocolConfig apn =
+            new ApplicationProtocolConfig(
+                ApplicationProtocolConfig.Protocol.ALPN,
+                ApplicationProtocolConfig.SelectorFailureBehavior.NO_ADVERTISE,
+                ApplicationProtocolConfig.SelectedListenerFailureBehavior.ACCEPT,
+                alpnList);
+        sslContextBuilder.applicationProtocolConfig(apn);
+      }
+      SslContext sslContextCopy = sslContextBuilder.build();
+      sslContext = sslContextCopy;
```

Comment:
How is `sslContext` synchronized between `updateSslContext ()` and `addCallback()`? Are `updateSslContext ()` and `addCallback()` run in the same thread or they are executed sequentially?

Reply:
No, they don't have to be sequential or from the same thread. I guess I could have used `AtomicReference` because all I need is atomic accesses to [sslContext](https://github.com/grpc/grpc-java/blob/2cf81bc5072fde5062d1738c2a2f336f602b389f/xds/src/main/java/io/grpc/xds/internal/sds/DynamicSslContextProvider.java#L41) but instead I used `sslContextCopy` (local copy) in both `updateSslContext ()` and `addCallback()` like this:
```
    SslContext sslContextCopy = sslContext;
   ... use sslContextCopy from here...
```
because [reference assignments are atomic](https://docs.oracle.com/javase/tutorial/essential/concurrency/atomic.html).

Reply:
There could still be data race:
(1) `addCallback()` checks `sslContext`, and it's `null`.
(2) then `updateSslContext ()` set `sslContext` and drains `penddingCallbacks`
(3) then `addCallback()` adds a callback to `penddingCallbacks` and it's never performed.

Reply:
Hmmm. Good catch. Basically the updates (and checks) to sslContext *and* pendingCallbacks should be combined as one atomic operation. I'll use pendingCallbacks as the lock. e.g. in addCallback:
```
synchronized (pendingCallbacks) {
    if (sslContext != null) {
      callPerformCallback(callback, executor, sslContext);
    } else {
      pendingCallbacks.add(new CallbackPair(callback, executor));
    }
}
```
Similarly in `updateSslContext`